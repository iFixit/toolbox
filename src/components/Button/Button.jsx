import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import constants from '../../constants';

const {
   borderRadius,
   color,
   fontSize,
   lineHeight,
   spacing,
   transition,
} = constants;

const propTypes = {
   /** Contents of the button. */
   children: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
      PropTypes.arrayOf(PropTypes.node),
   ]).isRequired,
   /** Choose a button design. */
   design: PropTypes.oneOf([
      'default',
      'primary',
      'secondary',
      'outline',
      'plain',
   ]),
   /** Disable button. */
   disabled: PropTypes.bool,
   /** Set width of the button to 100%. */
   fullWidth: PropTypes.bool,
   /** If an `href` is provided, the underlying element will be an `<a>` instead of a `<button>`. */
   href: PropTypes.string,
   /** Size of button. */
   size: PropTypes.oneOf(['small', 'medium', 'large']),
   /** Callback when clicked. */
   onClick: PropTypes.func,
};

const defaultProps = {
   design: 'default',
   disabled: false,
   fullWidth: false,
   href: '',
   size: 'medium',
   onClick: () => {},
};

const designs = {
   default: {
      color: color.grayAlpha[9],
      backgroundColor: 'transparent',
      borderColor: color.grayAlpha[3],
      ':hover': {
         backgroundColor: color.grayAlpha[1],
      },
      ':focus': {
         boxShadow: `0 0 0 3px ${color.grayAlpha[3]}`,
      },
   },
   primary: {
      color: color.white,
      backgroundColor: color.blue[4],
      borderColor: color.blue[4],
      ':hover': {
         color: color.white,
         backgroundColor: color.blue[5],
         borderColor: color.blue[5],
      },
      ':focus': {
         boxShadow: `0 0 0 3px ${color.blueAlpha[2]}`,
      },
   },
   secondary: {
      color: color.grayAlpha[9],
      backgroundColor: color.grayAlpha[3],
      ':hover': {
         backgroundColor: color.grayAlpha[4],
      },
      ':focus': {
         boxShadow: `0 0 0 3px ${color.grayAlpha[4]}`,
      },
   },
   outline: {
      color: color.blue[4],
      backgroundColor: 'transparent',
      borderColor: color.blue[4],
      ':hover': {
         color: color.white,
         backgroundColor: color.blue[4],
         borderColor: color.blue[4],
      },
      ':focus': {
         boxShadow: `0 0 0 3px ${color.blueAlpha[2]}`,
      },
   },
   plain: {
      color: color.grayAlpha[9],
      backgroundColor: 'transparent',
      borderColor: 'transparent',
      ':hover': {
         color: color.blue[4],
      },
      ':focus': {
         backgroundColor: color.grayAlpha[1],
      },
   },
};

const sizes = {
   small: {
      padding: `${spacing[0]} ${spacing[2]}`,
      fontSize: fontSize[1],
      lineHeight: lineHeight.copy,
   },
   medium: {
      padding: `${spacing[1]} ${spacing[3]}`,
      fontSize: fontSize[1],
      lineHeight: lineHeight.copy,
   },
   large: {
      padding: `${spacing[3]} ${spacing[4]}`,
      fontSize: fontSize[2],
      lineHeight: lineHeight.title,
   },
};

const ButtonContainer = styled.button(
   {
      display: 'inline-block',
      boxSizing: 'border-box',
      fontFamily: 'inherit',
      textDecoration: 'none',
      border: '1px solid transparent',
      borderRadius,
      outline: 0,
      whiteSpace: 'nowrap',
      cursor: 'pointer',
      userSelect: 'none',
      transition: `all ${transition.duration} ${transition.easing}`,
   },
   ({ design }) => designs[design],
   ({ size }) => sizes[size],
   ({ fullWidth }) =>
      fullWidth && {
         display: 'flex',
         width: '100%',
      },
   ({ disabled }) =>
      disabled && {
         opacity: 0.5,
         pointerEvents: 'none',
      },
);

const LinkContainer = ButtonContainer.withComponent('a');

// Buttons cannot be flex containers in some browsers (i.e. Safari).
// To get around this, we wrap the contents of the button in a flex container.
const FlexContainer = styled.span({
   width: '100%',
   display: 'flex',
   justifyContent: 'center',
   alignItems: 'center',
   textAlign: 'center',

   '> :not(:first-child)': {
      marginLeft: spacing[1],
   },
});

const TruncateContainer = styled.span({
   textOverflow: 'ellipsis',
   overflow: 'hidden',
   whiteSpace: 'nowrap',
});

const Button = ({ children, ...props }) => {
   const Container = props.href === '' ? ButtonContainer : LinkContainer;

   return (
      <Container {...props} tabIndex={props.disabled ? -1 : 0}>
         <FlexContainer>
            {React.Children.map(
               children,
               child =>
                  typeof child !== 'string' ? (
                     child
                  ) : (
                     <TruncateContainer>{child}</TruncateContainer>
                  ),
            )}
         </FlexContainer>
      </Container>
   );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
