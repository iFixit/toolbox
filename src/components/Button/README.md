### Simple Button

```
<Button>Button</Button>
```

### Button Designs

```
<Button>Default</Button>
```

```
<Button design="primary">Primary</Button>
```

```
<Button design="outline">Outline</Button>
```

### Button Sizes

```
<Button size="small">Small</Button>
```

```
<Button>Medium</Button>
```

```
<Button size="large">Large</Button>
```

### Full-Width Button

```
<Button fullWidth>Full-width</Button>
```

### Disabled Button

```
<Button disabled>Disabled</Button>
```

### onClick Callback

```
<Button onClick={() => console.log('clicked')}>Click me</Button>
```

### Link

```
<Button link href="https://ifixit.com" target="_blank">Go to iFixit.com</Button>
```
