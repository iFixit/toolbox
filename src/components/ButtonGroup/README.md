### Examples

#### Row

```
<ButtonGroup>
   <Button>File</Button>
   <Button>Edit</Button>
   <Button>View</Button>
   <Button>History</Button>
</ButtonGroup>
```

#### Column

```
<ButtonGroup direction="column">
   <Button>File</Button>
   <Button>Edit</Button>
   <Button>View</Button>
   <Button>History</Button>
</ButtonGroup>
```
