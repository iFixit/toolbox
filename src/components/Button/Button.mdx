---
name: Button
route: /components/button
menu: Components
---

import Button from './Button'
import Icon from '../Icon/Icon'
import { Playground, PropsTable } from 'docz'

# Button

## Examples

### Default

<Playground>
  <Button>Button</Button>
</Playground>

### Button designs

<Playground>
  <Button design="default">Default</Button>
</Playground>

<Playground>
  <Button design="primary">Primary</Button>
</Playground>

<Playground>
  <Button design="secondary">Secondary</Button>
</Playground>

<Playground>
  <Button design="outline">Outline</Button>
</Playground>

<Playground>
  <Button design="plain">Plain</Button>
</Playground>

### Icon buttons

<Playground>
  <Button>
    <Icon name="heart" size={16} />
    <span>Icon left</span>
  </Button>
</Playground>

<Playground>
  <Button>
    <span>Icon right</span>
    <Icon name="search" size={16} />
  </Button>
</Playground>

### Button sizes

<Playground>
  <Button size="small">Small</Button>
</Playground>

<Playground>
  <Button size="medium">Medium</Button>
</Playground>

<Playground>
  <Button size="large">Large</Button>
</Playground>

### Full-width Button

<Playground>
  <Button fullWidth>Full-width</Button>
</Playground>

### Disabled Button

<Playground>
  <Button disabled>Disabled</Button>
</Playground>

### onClick callback

<Playground>
  <Button onClick={() => console.log('clicked')}>Click me</Button>
</Playground>

### Link

<Playground>
  <Button href="https://ifixit.com" target="_blank">Go to iFixit.com</Button>
</Playground>

## Props

<PropsTable of={Button} />
