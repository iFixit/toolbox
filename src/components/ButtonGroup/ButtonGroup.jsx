import React from 'react';
import { arrayOf, bool, node } from 'prop-types';
import glamorous from 'glamorous';

import Button from '../Button/Button';
import constants from '../../constants';

const { borderRadius } = constants;
const { Div } = glamorous;

const ButtonGroupItem = glamorous(Button)({
   borderRadius: 0,
   borderRightWidth: 0,

   '&:first-of-type': {
      borderTopLeftRadius: borderRadius,
      borderBottomLeftRadius: borderRadius,
   },
   '&:last-of-type': {
      borderTopRightRadius: borderRadius,
      borderBottomRightRadius: borderRadius,
      borderRightWidth: 1,
   },
});

const ButtonGroup = ({ children, disabled, ...props }) => (
   <Div display="inline-block" {...props}>
      {React.Children.map(children, child => {
         if (child.type !== Button) {
            return child;
         }
         return <ButtonGroupItem {...child.props} disabled={disabled} />;
      })}
   </Div>
);

ButtonGroup.propTypes = {
   /** Buttons to be grouped together. */
   children: arrayOf(node).isRequired,
   /** Indicates that the control is not available for interaction */
   disabled: bool,
};

ButtonGroup.defaultProps = {
   disabled: false,
};

export default ButtonGroup;
