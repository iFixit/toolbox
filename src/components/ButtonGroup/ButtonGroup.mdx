---
name: ButtonGroup
route: /components/button-group
menu: Components
---

import ButtonGroup from './ButtonGroup'
import Button from '../Button/Button'
import { Playground, PropsTable } from 'docz'

# ButtonGroup

## Examples

### Row

<Playground>
  <ButtonGroup>
    <Button>File</Button>
    <Button>Edit</Button>
    <Button>View</Button>
    <Button>History</Button>
  </ButtonGroup>
</Playground>

### Column

<Playground>
  <ButtonGroup direction="column">
    <Button>File</Button>
    <Button>Edit</Button>
    <Button>View</Button>
    <Button>History</Button>
  </ButtonGroup>
</Playground>

## Props

<PropsTable of={ButtonGroup} />
