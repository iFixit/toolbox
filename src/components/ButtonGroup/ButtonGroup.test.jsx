import React from 'react';
import renderer from 'react-test-renderer';
import ButtonGroup from './ButtonGroup';
import Button from '../Button/Button';

test('renders without crashing', () => {
   const tree = renderer.create(<ButtonGroup />).toJSON();
   expect(tree).toMatchSnapshot();
});

test('renders when disabled', () => {
   const tree = renderer
      .create(
         <ButtonGroup disabled>
            <Button>Button 1</Button>
            <Button>Button 2</Button>
         </ButtonGroup>,
      )
      .toJSON();
   expect(tree).toMatchSnapshot();
});
