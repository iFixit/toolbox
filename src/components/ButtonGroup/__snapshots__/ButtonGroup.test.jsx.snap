// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when disabled 1`] = `
.glamor-6,
[data-glamor-6] {
  display: inline-block;
}

.glamor-1,
[data-glamor-1] {
  width: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.glamor-1> :not(:first-child),
[data-glamor-1]> :not(:first-child) {
  margin-left: 0.5rem;
}

.glamor-0,
[data-glamor-0] {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.glamor-2,
[data-glamor-2] {
  display: inline-block;
  box-sizing: border-box;
  font-family: inherit;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 0;
  outline: 0;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: rgba(0, 3, 6, 0.87);
  background-color: transparent;
  border-color: rgba(0, 3, 6, 0.12);
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  line-height: 1.5;
  opacity: 0.5;
  pointer-events: none;
  border-right-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.glamor-2:hover,
[data-glamor-2]:hover {
  background-color: rgba(0, 3, 6, 0.04);
}

.glamor-2:focus,
[data-glamor-2]:focus {
  box-shadow: 0 0 0 3px rgba(0, 3, 6, 0.12);
}

.glamor-2:first-of-type,
[data-glamor-2]:first-of-type {
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}

.glamor-2:last-of-type,
[data-glamor-2]:last-of-type {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-right-width: 1px;
}

<div
  className="glamor-6"
>
  <button
    className="glamor-2"
    disabled={true}
    onClick={[Function]}
    tabIndex={-1}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 1
      </span>
    </span>
  </button>
  <button
    className="glamor-2"
    disabled={true}
    onClick={[Function]}
    tabIndex={-1}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 2
      </span>
    </span>
  </button>
</div>
`;

exports[`renders with disabled children 1`] = `
.glamor-6,
[data-glamor-6] {
  display: inline-block;
}

.glamor-5,
[data-glamor-5] {
  display: inline-block;
  box-sizing: border-box;
  font-family: inherit;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 0;
  outline: 0;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: rgba(0, 3, 6, 0.87);
  background-color: transparent;
  border-color: rgba(0, 3, 6, 0.12);
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  line-height: 1.5;
  border-right-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.glamor-5:hover,
[data-glamor-5]:hover {
  background-color: rgba(0, 3, 6, 0.04);
}

.glamor-5:focus,
[data-glamor-5]:focus {
  box-shadow: 0 0 0 3px rgba(0, 3, 6, 0.12);
}

.glamor-5:first-of-type,
[data-glamor-5]:first-of-type {
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}

.glamor-5:last-of-type,
[data-glamor-5]:last-of-type {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-right-width: 1px;
}

.glamor-1,
[data-glamor-1] {
  width: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.glamor-1> :not(:first-child),
[data-glamor-1]> :not(:first-child) {
  margin-left: 0.5rem;
}

.glamor-0,
[data-glamor-0] {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.glamor-2,
[data-glamor-2] {
  display: inline-block;
  box-sizing: border-box;
  font-family: inherit;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 0;
  outline: 0;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: rgba(0, 3, 6, 0.87);
  background-color: transparent;
  border-color: rgba(0, 3, 6, 0.12);
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  line-height: 1.5;
  opacity: 0.5;
  pointer-events: none;
  border-right-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.glamor-2:hover,
[data-glamor-2]:hover {
  background-color: rgba(0, 3, 6, 0.04);
}

.glamor-2:focus,
[data-glamor-2]:focus {
  box-shadow: 0 0 0 3px rgba(0, 3, 6, 0.12);
}

.glamor-2:first-of-type,
[data-glamor-2]:first-of-type {
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}

.glamor-2:last-of-type,
[data-glamor-2]:last-of-type {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-right-width: 1px;
}

<div
  className="glamor-6"
>
  <button
    className="glamor-2"
    disabled={true}
    onClick={[Function]}
    tabIndex={-1}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 1
      </span>
    </span>
  </button>
  <button
    className="glamor-5"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 2
      </span>
    </span>
  </button>
</div>
`;

exports[`renders with non-button children 1`] = `
.glamor-6,
[data-glamor-6] {
  display: inline-block;
}

.glamor-2,
[data-glamor-2] {
  display: inline-block;
  box-sizing: border-box;
  font-family: inherit;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 0;
  outline: 0;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: rgba(0, 3, 6, 0.87);
  background-color: transparent;
  border-color: rgba(0, 3, 6, 0.12);
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  line-height: 1.5;
  border-right-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.glamor-2:hover,
[data-glamor-2]:hover {
  background-color: rgba(0, 3, 6, 0.04);
}

.glamor-2:focus,
[data-glamor-2]:focus {
  box-shadow: 0 0 0 3px rgba(0, 3, 6, 0.12);
}

.glamor-2:first-of-type,
[data-glamor-2]:first-of-type {
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}

.glamor-2:last-of-type,
[data-glamor-2]:last-of-type {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-right-width: 1px;
}

.glamor-1,
[data-glamor-1] {
  width: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.glamor-1> :not(:first-child),
[data-glamor-1]> :not(:first-child) {
  margin-left: 0.5rem;
}

.glamor-0,
[data-glamor-0] {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div
  className="glamor-6"
>
  <button
    className="glamor-2"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 1
      </span>
    </span>
  </button>
  <button
    className="glamor-2"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 2
      </span>
    </span>
  </button>
  <span>
    Span
  </span>
</div>
`;

exports[`renders without crashing 1`] = `
.glamor-6,
[data-glamor-6] {
  display: inline-block;
}

.glamor-2,
[data-glamor-2] {
  display: inline-block;
  box-sizing: border-box;
  font-family: inherit;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 0;
  outline: 0;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: rgba(0, 3, 6, 0.87);
  background-color: transparent;
  border-color: rgba(0, 3, 6, 0.12);
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  line-height: 1.5;
  border-right-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.glamor-2:hover,
[data-glamor-2]:hover {
  background-color: rgba(0, 3, 6, 0.04);
}

.glamor-2:focus,
[data-glamor-2]:focus {
  box-shadow: 0 0 0 3px rgba(0, 3, 6, 0.12);
}

.glamor-2:first-of-type,
[data-glamor-2]:first-of-type {
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}

.glamor-2:last-of-type,
[data-glamor-2]:last-of-type {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-right-width: 1px;
}

.glamor-1,
[data-glamor-1] {
  width: 100%;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.glamor-1> :not(:first-child),
[data-glamor-1]> :not(:first-child) {
  margin-left: 0.5rem;
}

.glamor-0,
[data-glamor-0] {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div
  className="glamor-6"
>
  <button
    className="glamor-2"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 1
      </span>
    </span>
  </button>
  <button
    className="glamor-2"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
  >
    <span
      className="glamor-1"
    >
      <span
        className="glamor-0"
      >
        Button 2
      </span>
    </span>
  </button>
</div>
`;
